src/sema/expression/function_call.rs

```diff
@@ -2179,6 +2179,21 @@ pub(super) fn parse_call_args(
                     .to_string(),
             ));
             return Err(());
+        } else if res.accounts.is_none()
+            && !matches!(
+                ns.functions[context.function_no.unwrap()].visibility,
+                Visibility::External(_)
+            )
+            && !ns.functions[context.function_no.unwrap()].is_constructor()
+        {
+            std::println!("{}", ns.functions[context.function_no.unwrap()].visibility);
+            diagnostics.push(Diagnostic::error(
+                *loc,
+                "A constructor call in a non-external function needs the AccountMeta array. \
+                The compiler can only automatically handle accounts in functions declared as external"
```

Comment:
I don't think this is clear. It is written from the perspective of the compiler writer, not the Solidity developer

Accounts are required for calling a contract constructor. You can either provide the accounts with the {accounts: [ ...]} call arguments or change the visibility of this function to external.